pkg/deployment/resources/servicemonitor.go

```diff
@@ -0,0 +1,157 @@
+// and keeps it in the Resources.
+func (r *Resources) EnsureMonitoringClient() (*clientv1.MonitoringV1Client, error) {
+	if r.monitoringClient != nil {
+		return r.monitoringClient, nil
+	}
+
+	// Make a client:
+	var restConfig *rest.Config
+	restConfig, err := k8sutil.InClusterConfig()
+	if err != nil {
+		return nil, maskAny(err)
+	}
+	mClient, err := clientv1.NewForConfig(restConfig)
+	if err != nil {
+		return nil, maskAny(err)
+	}
+	r.monitoringClient = mClient
+	return mClient, nil
+}
+
+func (r *Resources) makeEndpoint(isSecure bool) coreosv1.Endpoint {
+	if isSecure {
+		return coreosv1.Endpoint{
+			Port:     "exporter",
+			Interval: "10s",
+			Scheme:   "https",
+			TLSConfig: &coreosv1.TLSConfig{
+				InsecureSkipVerify: true,
+			},
+		}
+	} else {
+		return coreosv1.Endpoint{
+			Port:     "exporter",
+			Interval: "10s",
+			Scheme:   "http",
+		}
+	}
+}
+
+// EnsureServiceMonitor creates or updates a ServiceMonitor.
+func (r *Resources) EnsureServiceMonitor() error {
+	// Some preparations:
+	log := r.log
+	apiObject := r.context.GetAPIObject()
+	deploymentName := apiObject.GetName()
+	ns := apiObject.GetNamespace()
+	owner := apiObject.AsOwner()
+	spec := r.context.GetSpec()
+	wantMetrics := spec.Metrics.IsEnabled()
+	serviceMonitorName := k8sutil.CreateExporterClientServiceName(deploymentName)
+
+	mClient, err := r.EnsureMonitoringClient()
+	if err != nil {
+		log.Error().Err(err).Msgf("Cannot get a monitoring client.")
+		return maskAny(err)
+	}
+
+	// Check if ServiceMonitor already exists
+	serviceMonitors := mClient.ServiceMonitors(ns)
+	_, err = serviceMonitors.Get(serviceMonitorName, metav1.GetOptions{})
+	if err != nil {
+		if k8sutil.IsNotFound(err) {
+			if !wantMetrics {
+				return nil
+			}
+			// Need to create one:
+			smon := &coreosv1.ServiceMonitor{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:            serviceMonitorName,
+					Labels:          LabelsForExporterServiceMonitor(deploymentName),
+					OwnerReferences: []metav1.OwnerReference{owner},
+				},
+				Spec: coreosv1.ServiceMonitorSpec{
+					JobLabel: "k8s-app",
+					Endpoints: []coreosv1.Endpoint{
+						r.makeEndpoint(spec.IsSecure()),
+					},
+					Selector: metav1.LabelSelector{
+						MatchLabels: LabelsForExporterServiceMonitorSelector(deploymentName),
+					},
+				},
+			}
+			smon, err = serviceMonitors.Create(smon)
+			if err != nil {
+				log.Error().Err(err).Msgf("Failed to create ServiceMonitor %s", serviceMonitorName)
+				return maskAny(err)
+			}
+			log.Debug().Msgf("ServiceMonitor %s successfully created.", serviceMonitorName)
+			return nil
+		} else {
+			log.Error().Err(err).Msgf("Failed to get ServiceMonitor %s", serviceMonitorName)
+			return maskAny(err)
+		}
+	}
+	if wantMetrics {
+		log.Debug().Msgf("ServiceMonitor %s already found, no need to create.",
+			serviceMonitorName)
+		return nil
+	}
+	// Need to get rid of the ServiceMonitor:
+	err = serviceMonitors.Delete(serviceMonitorName, &metav1.DeleteOptions{})
```

Comment:
Please check that the apiObject owns the monitor before deleting it.

Reply:
Ping on this!

Reply:
Right.

Reply:
Done.